compiler/noirc_frontend/src/hir/comptime/interpreter.rs

```diff
@@ -895,71 +897,155 @@ impl<'local, 'interner> Interpreter<'local, 'interner> {
     }
 
     fn evaluate_infix(&mut self, infix: HirInfixExpression, id: ExprId) -> IResult<Value> {
-        let lhs = self.evaluate(infix.lhs)?;
-        let rhs = self.evaluate(infix.rhs)?;
+        let lhs_value = self.evaluate(infix.lhs)?;
+        let rhs_value = self.evaluate(infix.rhs)?;
 
         if self.elaborator.interner.get_selected_impl_for_expression(id).is_some() {
-            return self.evaluate_overloaded_infix(infix, lhs, rhs, id);
+            return self.evaluate_overloaded_infix(infix, lhs_value, rhs_value, id);
         }
 
-        let make_error = |this: &mut Self, lhs: Value, rhs: Value, operator| {
-            let location = this.elaborator.interner.expr_location(&id);
+        fn make_error_value(
+            lhs: &Value,
+            rhs: &Value,
+            location: Location,
+            operator: &'static str,
+        ) -> InterpreterError {
             let lhs = lhs.get_type().into_owned();
             let rhs = rhs.get_type().into_owned();
-            Err(InvalidValuesForBinary { lhs, rhs, location, operator })
+            InterpreterError::InvalidValuesForBinary { lhs, rhs, location, operator }
+        }
+
+        // make_error_value if the result is None
+        fn or_err<T>(
+            lhs_rhs_location: (&Value, &Value, Location),
+            operator: &'static str,
+            result: Option<T>,
+        ) -> IResult<T> {
+            let (lhs, rhs, location) = lhs_rhs_location;
+            result.ok_or(make_error_value(lhs, rhs, location, operator))
+        }
+
+        let location = self.elaborator.interner.expr_location(&id);
+        let error_ctx = (&lhs_value, &rhs_value, location);
+        let make_error = |lhs: Value, rhs: Value, operator| -> IResult<Value> {
+            Err(make_error_value(&lhs, &rhs, location, operator))
```

Comment:
We can replace this setup along with `make_error_value` with
```rs
        let lhs_type = lhs_value.get_type().into_owned();
        let rhs_type = rhs_value.get_type().into_owned();
        let location = self.elaborator.interner.expr_location(&id);

        let error = |operator| {
            let lhs = lhs_type;
            let rhs = rhs_type;
            InterpreterError::InvalidValuesForBinary { lhs, rhs, location, operator }
        };
```
Then each match case becomes
```rs
                (Value::I8(lhs), Value::I8(rhs)) => {
                    Ok(Value::I8(lhs.checked_add(rhs).ok_or(error("+"))?))
                }
```
and the catch-all make_error cases are now:
```rs
                _ => Err(error("+")),
```

Reply:
That worked with one small change: 
```
            let lhs = lhs_type.clone();
            let rhs = rhs_type.clone();
```
otherwise the double-usage when `try_into()` is called throws an error